### PR TITLE
Calculating and updtating cluster and volume alert_count in each cluster sync

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
tendrl-bug-id: Tendrl/gluster-integration#598

Signed-off-by: GowthamShanmugasundaram <gshanmug@redhat.com>

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
@shtripat @shirshendu @r0h4n please review

**Comment by GowthamShanmugam** [comment]
Tendrl/monitoring-integration/issues/373

**Comment by GowthamShanmugam** [comment]
only provisioning node will do alert_count update

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 1]
```diff
@@ -862,3 +820,59 @@ def brick_status_alert(hostname):
...
+        logger.log(
+            "debug",
+            NS.publisher_id,
+            {"message": "Unable to update alert count.err: %s" % ex}
+        )
+
+
+def find_volume_id():
+    alert_counts = {}
+    volumes = etcd_utils.read(
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: Why not use `NS.gluster.objects.Volume().load_all()` here ? I feel all you need is volume names then form the dict.
Also call the function as `get_volume_alert_counts`

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 2]
```diff
@@ -862,3 +820,59 @@ def brick_status_alert(hostname):
...
         if isinstance(lock, etcd.lock.Lock) and lock.is_acquired:
             lock.release()
+
+
+def update_cluster_alert_count():
+    cluster_alert_count = 0
+    severity = ["WARNING", "CRITICAL"]
+    try:
+        alert_counts = find_volume_id()
+        alerts_arr = NS.tendrl.objects.ClusterAlert(
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: Is it possible to load all the cluster alerts using tags only? Don't you need integration_id here? Does it default to NS.tendrl_context.integration_id in object constructor?

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 3]
```diff
@@ -862,3 +820,59 @@ def brick_status_alert(hostname):
...
+            tags={'integration_id': NS.tendrl_context.integration_id}
+        ).load_all()
+        for alert in alerts_arr:
+            alert.tags = json.loads(alert.tags)
+            if alert.severity in severity:
+                cluster_alert_count += 1
+                if alert.resource in NS.gluster.objects.VolumeAlertCounters(
+                        )._defs['relationship'][alert.alert_type.lower()]:
+                    vol_name = alert.tags.get('volume_name', None)
+                    if vol_name:
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: Add both the conditions in one if condition as `if vol_name and vol_name in alert_counts.keys()`

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 4]
```diff
@@ -862,3 +820,59 @@ def brick_status_alert(hostname):
...
+                        if vol_name in alert_counts.keys():
+                            alert_counts[vol_name]['alert_count'] += 1
+        # Update cluster alert count
+        NS.tendrl.objects.ClusterAlertCounters(
+            integration_id=NS.tendrl_context.integration_id,
+            alert_count=cluster_alert_count
+        ).save()
+        # Update volume alert count
+        for volume in alert_counts:
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: I understand alert_counts is a dict. If so you should use something like

```
for vol_name, alert_det in alert_counts.iter_items():
    NS.gluster.objects.VolumeAlertCounters(
        integration_id=NS.tendrl_context.integration_id,
        alert_count=alert_det['alert_count'],
        volume_id=alert_det['vol_id']
    ).save()
```

Does this work. Even I feel the dict created in `find_volume_id` functions could be with keys as volume ids and value as counters. Why you need volume name as keys? Do you really use that anywhere?

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 5]
```diff
@@ -862,3 +820,59 @@ def brick_status_alert(hostname):
...
+            "debug",
+            NS.publisher_id,
+            {"message": "Unable to update alert count.err: %s" % ex}
+        )
+
+
+def find_volume_id():
+    alert_counts = {}
+    volumes = etcd_utils.read(
+        "clusters/%s/Volumes" % NS.tendrl_context.integration_id
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: i will change this

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 1 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: ok

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 2 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: @shtripat for the cluster_alert object is a subclass of alert object, so for an alert object integration_id is not a mandatory field. So for cluster alerts, integration_id is present in tags only. While forming patch cluster alert object will take integration_id from tags only. That is why i am passing only tags with integration_id.

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 3 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: i will change this

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 4 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: the alert object contains volume name only present in tags, to compare this alert with a particular volume i need volume name. that why i am creating dict with the volume name. During alert iteration, i am just comparing volume name from alert object with newly formed dict to increase the alert count.

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 2 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: https://github.com/Tendrl/commons/blob/master/tendrl/commons/objects/cluster_alert/__init__.py

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 1 above)
Comment (GowthamShanmugam) [reply tendrl/gluster_integration/sds_sync/__init__.py]: i am just reading, volume_id and forming a dict with alert_count is 0, how can we call this function get_volume_alert_counts?

**Comment by GowthamShanmugam** [comment]
@shtripat changes done

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 2 above)
Comment (shtripat) [reply tendrl/gluster_integration/sds_sync/__init__.py]: ok

File: tendrl/gluster_integration/sds_sync/__init__.py (hunk 4 above)
Comment (shtripat) [reply tendrl/gluster_integration/sds_sync/__init__.py]: ok

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 6]
```diff
@@ -837,42 +837,33 @@ def update_cluster_alert_count():
...
-            key = volume.key + "/name"
-            vol_name = etcd_utils.read(key).value
-            alert_counts[vol_name] = {}
-            alert_counts[vol_name]['vol_id'] = volume_id
-            alert_counts[vol_name]['alert_count'] = 0
-        except etcd.EtcdKeyNotFound:
-            continue
+    volumes = NS.gluster.objects.Volume().load_all()
+    for volume in volumes:
+        alert_counts[volume.name] = {}
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: you can do the whole thing in one line 

`alert_counts[volume.name] = {"vol_id": volume.vol_id, "alert_count": 0}`

**Comment by GowthamShanmugam** [comment]
@shtripat changes done

File: tendrl/gluster_integration/sds_sync/__init__.py [hunk 7]
```diff
@@ -862,3 +820,50 @@ def brick_status_alert(hostname):
...
         if isinstance(lock, etcd.lock.Lock) and lock.is_acquired:
             lock.release()
+
+
+def update_cluster_alert_count():
+    cluster_alert_count = 0
+    severity = ["WARNING", "CRITICAL"]
+    try:
+        alert_counts = get_volume_alert_counts()
+        alerts_arr = NS.tendrl.objects.ClusterAlert(
```
Comment (shtripat) [inline tendrl/gluster_integration/sds_sync/__init__.py]: `alerts` is good enough name :)

**Comment by shtripat** [comment]
Looks fine. One small suggestion.
Also have you verified these changes?

**Comment by GowthamShanmugam** [comment]
@shtripat changes done, i verified

**Comment by GowthamShanmugam** [comment]
Tendrl/ui/issues#867

**Pull request merged**